src/coreclr/jit/codegenxarch.cpp

```diff
@@ -1010,7 +1010,23 @@ void CodeGen::genCodeForBinary(GenTreeOp* treeNode)
     // the same code as above
     else if (op2reg == targetReg)
     {
-        noway_assert(GenTree::OperIsCommutative(oper));
+        unsigned lclNum1 = (unsigned)-1;
+        unsigned lclNum2 = (unsigned)-2;
+
+#ifdef DEBUG
+        GenTree* op1Skip = op1->gtSkipReloadOrCopy();
+        GenTree* op2Skip = op2->gtSkipReloadOrCopy();
+
+        if (op1Skip->OperIsLocalRead())
+        {
+            lclNum1 = op1Skip->AsLclVarCommon()->GetLclNum();
+        }
+        if (op2Skip->OperIsLocalRead())
+        {
+            lclNum2 = op2Skip->AsLclVarCommon()->GetLclNum();
+        }
+#endif
+        noway_assert(GenTree::OperIsCommutative(oper) || (lclNum1 == lclNum2));
```

Comment:
I realized this can't be a `noway_assert` now: it would mean we're asserting different things in Release and Debug: and the case where you're asserting `lclNum1 == lclNum2` would fail in Release.